### PR TITLE
Add BigQuery UDF support

### LINKED ISSUE #1829: BigQuery UDF is not usable?
Hi,
I want to use UDF in the query of BigQuery.
gcloud-python supports it?

### PR DESCRIPTION
**Pull request by takbab**
Add a field to support BigQuery UDF.
https://cloud.google.com/bigquery/docs/reference/v2/jobs
- configuration.query.userDefinedFunctionResources[]
- configuration.query.userDefinedFunctionResources[].inlineCode
- configuration.query.userDefinedFunctionResources[].resourceUri

Closes #1829

### CODE REVIEW

File: gcloud/bigquery/job.py [hunk 1]
```diff
@@ -927,6 +928,11 @@ def __init__(self, name, query, client):
     https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.query.useQueryCache
     """
 
+    user_defined_function_resources = _TypedProperty('user_defined_function_resources', list)
+    """See:
+    https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.query.userDefinedFunctionResources
+    """
```
Comment (tseaver) [inline gcloud/bigquery/job.py]: Using `_TypedProperty` for a list of mappings leaves us with no checking at assignment time.  @dhermes, do you think we should have a "setter" for the property which checks that the list elements are dicts with appropriate keys?

File: gcloud/bigquery/job.py [hunk 2]
```diff
@@ -927,6 +928,11 @@ def __init__(self, name, query, client):
     https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.query.useQueryCache
     """
 
+    user_defined_function_resources = _TypedProperty('user_defined_function_resources', list)
```
Comment (tseaver) [inline gcloud/bigquery/job.py]: pep8 doesn't like the long line.  Please wrap to < 79 characters (e.g., just after `_TypedProperty`).

File: gcloud/bigquery/job.py [hunk 3]
```diff
@@ -965,6 +971,8 @@ def _populate_config_resource(self, configuration):
             configuration['priority'] = self.priority
         if self.use_query_cache is not None:
             configuration['useQueryCache'] = self.use_query_cache
+        if self.user_defined_function_resources is not None:
+            configuration['userDefinedFunctionResources'] = self.user_defined_function_resources
```
Comment (tseaver) [inline gcloud/bigquery/job.py]: pep8 doesn't like the long line.  Please wrap to < 79 characters (e.g., open a parenthesis after the `=` and wrap to the next line).

**Comment by tseaver** [comment]
@takbab Thanks for pitching in on this feature!

**Comment by takbab** [comment]
@tseaver Thanks for checking!

File: gcloud/bigquery/job.py (hunk 2 above)
Comment (takbab) [inline gcloud/bigquery/job.py]: modified

File: gcloud/bigquery/job.py (hunk 3 above)
Comment (takbab) [inline gcloud/bigquery/job.py]: modified

File: gcloud/bigquery/job.py [hunk 4]
```diff
@@ -869,6 +869,7 @@ class _AsyncQueryConfiguration(object):
     _flatten_results = None
     _priority = None
     _use_query_cache = None
+    _user_defined_function_resources = None
```
Comment (tseaver) [inline gcloud/bigquery/job.py]: https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/145240341#L464

Per the pylint defaults (which we don't override), this class attribute name is too long:  it wants a maximum of 30 characters.  We could spell it `_udf_resources` (and change the property accordingly.

**Comment by tseaver** [comment]
@takbab The e-mail address associated with commit d3dcfcc is not recognized by Github as belonging to your account, and therefore the CLA bot is flagging that commit.

**Comment by takbab** [comment]
@tseaver I will recreate PR to clean my commits and diff.

**Pull request closed without merging**